src/rosdistro/freeze_source.py

```diff
@@ -0,0 +1,106 @@
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+import sys
+import subprocess
+import threading
+import time
+import yaml
+
+try:
+    import queue
+except ImportError:
+    import Queue as queue
+
+CONCURRENT_DEFAULT = 16
+
+
+def freeze_distribution_sources(dist, release_version=False, release_tag=False,
+                                concurrent_ops=CONCURRENT_DEFAULT, quiet=False):
+    # Populate this queue with tuples of repositories instances to be updated,
+    # so that this work can be spread across multiple threads.
+    work_queue = queue.Queue()
+    for repo_name, repo in dist.repositories.iteritems():
+        # Only manipulate distribution entries with a source repo listed.
+        if repo.source_repository:
+            # Decide which git ref string we'll be using as the replacement match.
+            if repo.release_repository and (release_version or release_tag):
+                version = repo.release_repository.version.split('-')[0]
+            else:
+                version = repo.source_repository.version
+            work_queue.put((repo.source_repository, version, release_tag))
+
+    total_items = work_queue.qsize()
+
+    for i in range(concurrent_ops):
+        threading.Thread(target=_worker, args=[work_queue]).start()
+
+    # Wait until the threads have done all the work and exited.
+    while not work_queue.empty():
+        time.sleep(0.1)
+        if not quiet:
+            sys.stdout.write("Updating source repo versions (%d/%d)   \r" %
+                             (total_items - work_queue.qsize(), total_items))
+            sys.stdout.flush()
+    work_queue.join()
+
+    # Clear past the updating line.
+    if not quiet:
+        print("")
+
+
+def _worker(work_queue):
+    while True:
+        try:
+            source_repo, freeze_version, freeze_to_tag = work_queue.get(block=False)
+            cmd = ['git', 'ls-remote', source_repo.url]
+            ls_remote_lines = subprocess.check_output(cmd).splitlines()
+            for line in ls_remote_lines:
+                hash, ref = line.split('\t', 1)
+                if ref.endswith(freeze_version):
+                    if freeze_to_tag and ref.startswith('refs/tags/'):
+                        source_repo.version = ref.split('refs/tags/')[1]
+                    else:
+                        source_repo.version = hash
+                    break
+
+            work_queue.task_done()
+
+        except subprocess.CalledProcessError:
+            sys.stderr.write("Non-zero return code for: %s" % ' '.join(cmd))
```

Comment:
This line lacks a trailing newline. Since it is very common to miss when using `stderr` directly I recommend using:

    from __future__ import print_function

    print('Non-zero return code for: %s' % ' '.join(cmd), file=sys.stderr)